streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerWithKey.java

```diff
@@ -75,14 +75,21 @@
     void init(final ProcessorContext context);
 
     /**
-     * Transform the given [key and ]value to a new value.
+     * Transform the given [key and] value to a new value.
      * Additionally, any {@link StateStore} that is {@link KStream#transformValues(ValueTransformerWithKeySupplier, String...)
      * attached} to this operator can be accessed and modified arbitrarily (cf.
      * {@link ProcessorContext#getStateStore(String)}).
      * <p>
-     * Note, that using {@link ProcessorContext#forward(Object, Object)} or
+     * Note that using {@link ProcessorContext#forward(Object, Object)} or
      * {@link ProcessorContext#forward(Object, Object, To)} is not allowed within {@code transform} and
      * will result in an {@link StreamsException exception}.
+     * <p>
+     * Note that if a {@code ValueTransformerWithKey} is used in a {@link KTable#transformValues(ValueTransformerWithKeySupplier, String...)}
+     * (or any other overload of {@code KTable#transformValues(...)}) operation,
+     * then the provided {@link ProcessorContext} from {@link #init(ProcessorContext)}
+     * does not guarantee that all context information will be available when {@code transform()}
+     * is executed, as it might be executed "out-of-band" due to some internal optimizations
+     * applied by the Kafka Streams DSL.
```

Comment:
@vvcephei Updated the text a little bit.

Don't think we should talk about punctuations because it does not apply for the "ValueGetter" optimization case that we fixed. Did a follow up PR instead: https://github.com/apache/kafka/pull/10810

Reply:
Thanks!